File: Forecasting/Airlines.R
library(readxl)
airlines <- read_xlsx(file.choose()) # read Airlines data
View(airlines) # Seasonlity 12 months

# pre processing 

# so creating 12 dummy variables

x<- data.frame(outer(rep(month.abb,length = 96), month.abb,"==") + 0 )# Creating dummies for 12 months
View(x)
colnames(x)<-month.abb # Assigning month names 
View(x)
airlinesPassengers <- cbind(airlines,x)
colnames(airlinesPassengers)

# input t
airlinesPassengers["t"] <- c(1:96)
View(airlinesPassengers)

airlinesPassengers["log_passengers"] <- log(airlinesPassengers["Passengers"])
airlinesPassengers["t_square"] <- airlinesPassengers["t"]*airlinesPassengers["t"]
View(airlinesPassengers)
# preprocessing completed

attach(airlinesPassengers)
# partitioning 

train <- airlinesPassengers[1:84,]
test <- airlinesPassengers[85:96,]


############ Linear Model ####################

linear_Model <- lm(Passengers ~ t, data = train)
summary(linear_Model)
liner_pred <- data.frame(predict(linear_Model, interval = 'predict', newdata = test))
rmse_linear <- sqrt(mean((test$Passengers - liner_pred$fit)^2, na.rm = T))
rmse_linear # 53.19924

########### Exponential  ##############

expo_model <- lm(log_passengers ~ t, data = train)
summary(expo_model)
expo_pred <- data.frame(predict(expo_model, interval = 'predict', newdata = test))
rmse_expo <- sqrt(mean((test$Passengers - expo_pred$fit)^2, na.rm = T))
rmse_expo # 325.7151

########### Quadratic  ##########################

Quad_model <- lm(Passengers ~ t + t_square, data = train)
summary(Quad_model)
Quad_pred <- data.frame(predict(Quad_model, interval = 'predict', newdata = test))
rmse_Quad <- sqrt(mean((test$Passengers - Quad_pred$fit)^2, na.rm = T))
rmse_Quad # 48.05189

 ######################### Additive Seasonality #########################

sea_add_model<-lm(Passengers ~Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train)
summary(sea_add_model)
sea_add_pred<-data.frame(predict(sea_add_model,newdata=test,interval='predict'))
rmse_sea_add<-sqrt(mean((test$Passengers-sea_add_pred$fit)^2,na.rm = T))
rmse_sea_add  # 132.8198

######################## Additive Seasonality with Linear #################

Add_sea_Linear_model<-lm(Passengers~t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train)
summary(Add_sea_Linear_model)
Add_sea_Linear_pred<-data.frame(predict(Add_sea_Linear_model,interval='predict',newdata=test))
rmse_Add_sea_Linear<-sqrt(mean((test$Passengers-Add_sea_Linear_pred$fit)^2,na.rm=T))
rmse_Add_sea_Linear #  35.34896

######################## Additive Seasonality with Quadratic #################

Add_sea_Quad_model<-lm(Passengers~t+t_square+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train)
summary(Add_sea_Quad_model)
Add_sea_Quad_pred<-data.frame(predict(Add_sea_Quad_model,interval='predict',newdata=test))
rmse_Add_sea_Quad<-sqrt(mean((test$Passengers-Add_sea_Quad_pred$fit)^2,na.rm=T))
rmse_Add_sea_Quad # 26.36082

######################### Multiplicative Seasonality #########################

multi_sea_model<-lm(log_passengers~Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data = train)
summary(multi_sea_model)
multi_sea_pred<-data.frame(predict(multi_sea_model,newdata=test,interval='predict'))
rmse_multi_sea<-sqrt(mean((test$Passengers-exp(multi_sea_pred$fit))^2,na.rm = T))
rmse_multi_sea # 140.0632

### preparing table on model and it's RMSE values

table_rmse <- data.frame(c("rmse_linear","rmse_expo","rmse_Quad","rmse_sea_add","rmse_Add_sea_Quad", "rmse_mulri_sea"),c(rmse_linear,rmse_expo,rmse_Quad,rmse_sea_add,rmse_Add_sea_Quad, rmse_multi_sea))
colnames(table_rmse) <- c("model","RMSE")
View(table_rmse)

# Additive seasonality with Quadratic has least RMSE value

write.csv(airlinesPassengers, file="airlinesPassengers.csv", row.names = F)
getwd()

### combing Training and Test data to build Additive Seasonality using Quadratic Tred #####
Add_sea_Quad_model_final <- lm(Passengers ~ t + t_square+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov, data = airlinesPassengers)
summary(Add_sea_Quad_model_final)

###### Predicting the new data #####

new_data <- read.csv(file.choose())
View(new_data)
pred_new <- data.frame(predict(Add_sea_Quad_model_final, newdata = new_data, interval = 'predict'))
View(pred_new)

# Plot (Add_sea_Quad_model_final)
acf(Add_sea_Quad_model_final$residuals, lag.max = 10) # take all residual value of the model

A <- arima(Add_sea_Quad_model_final$residuals, order = c(1,0,0))
A$residuals

ARerrors <- A$residuals

acf(ARerrors, lag.max = 10)

# predicting next 12 months errors using arima(order = c(1,0,0))

library(forecast)
errors_12 <- forecast(A, h = 12)

View(errors_12)
future_errors <- data.frame(errors_12)
View(future_errors)
class(future_errors)
future_errors <- future_errors$Point.Forecast

# predicated values for new data + future error values

predicted_new_values <- pred_new + future_errors

write.csv(predicted_new_values, file = "predicted_new_values.csv")
getwd()

File: Forecasting/Airlines.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
airlines = pd.read_excel("file:///D:/DATA SCIENCE/ExcelR/Assignments/Forecasting/Airlines+Data.xlsx")
airlines.head(10)
#month =['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'] 
#p = airlines["Month"][0]
#p[0:10]

from datetime import datetime, time
airlines.columns
airlines["Passengers"].plot()


airlines["month"] = airlines["Month"].dt.strftime("%b") # month extraction
airlines["year"] = airlines.Month.dt.strftime("%y") # year extraction

import seaborn as sns
heatmap_y_month = pd.pivot_table(data=airlines,values="Passengers",index="year",columns="month",aggfunc="mean",fill_value=0)
sns.heatmap(heatmap_y_month,annot=True,fmt="g")

# Boxplot for ever
sns.boxplot(x="month",y="Passengers",data=airlines)
sns.boxplot(x="year",y="Passengers",data=airlines)

# Line plot for Passengers based on year  and for each month
#sns.lineplot(x="year",y="Passengers",hue="month",data=airlines)

month_dummies = pd.DataFrame(pd.get_dummies(airlines['month']))
airlines1 = pd.concat([airlines,month_dummies],axis = 1)

airlines1["t"] = np.arange(1,97)

airlines1["t_squared"] = airlines1["t"]*airlines1["t"]
airlines1.columns
airlines1["log_Passenger"] = np.log(airlines1["Passengers"])
#airlines1.rename(columns={"Ridership ('000)": 'Ridership'}, inplace=True)
airlines1["Passengers"].plot()
Train = airlines1.head(80)
Test = airlines1.tail(16)

# to change the index value in pandas data frame 
# Test.set_index(np.arange(1,13))

####################### L I N E A R ##########################
import statsmodels.formula.api as smf 

linear_model = smf.ols('Passengers~t',data=Train).fit()
pred_linear =  pd.Series(linear_model.predict(pd.DataFrame(Test['t'])))
rmse_linear = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_linear))**2))
rmse_linear

##################### Exponential ##############################

Exp = smf.ols('log_Passenger~t',data=Train).fit()
pred_Exp = pd.Series(Exp.predict(pd.DataFrame(Test['t'])))
rmse_Exp = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Exp)))**2))
rmse_Exp

#################### Quadratic ###############################

Quad = smf.ols('Passengers~t+t_squared',data=Train).fit()
pred_Quad = pd.Series(Quad.predict(Test[["t","t_squared"]]))
rmse_Quad = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_Quad))**2))
rmse_Quad

################### Additive seasonality ########################

add_sea = smf.ols('Passengers~Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data=Train).fit()
pred_add_sea = pd.Series(add_sea.predict(Test[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov']]))
rmse_add_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_add_sea))**2))
rmse_add_sea

################## Additive Seasonality Quadratic ############################

add_sea_Quad = smf.ols('Passengers~t+t_squared+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data=Train).fit()
pred_add_sea_quad = pd.Series(add_sea_Quad.predict(Test[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','t','t_squared']]))
rmse_add_sea_quad = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_add_sea_quad))**2))
rmse_add_sea_quad 

################## Multiplicative Seasonality ##################

Mul_sea = smf.ols('log_Passenger~Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit()
pred_Mult_sea = pd.Series(Mul_sea.predict(Test))
rmse_Mult_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_sea)))**2))
rmse_Mult_sea

##################Multiplicative Additive Seasonality ###########

Mul_Add_sea = smf.ols('log_Passenger~t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit()
pred_Mult_add_sea = pd.Series(Mul_Add_sea.predict(Test))
rmse_Mult_add_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_add_sea)))**2))
rmse_Mult_add_sea 

################## Testing #######################################

data = {"MODEL":pd.Series(["rmse_linear","rmse_Exp","rmse_Quad","rmse_add_sea","rmse_add_sea_quad","rmse_Mult_sea","rmse_Mult_add_sea"]),"RMSE_Values":pd.Series([rmse_linear,rmse_Exp,rmse_Quad,rmse_add_sea,rmse_add_sea_quad,rmse_Mult_sea,rmse_Mult_add_sea])}
table_rmse=pd.DataFrame(data)
table_rmse
# so rmse_Mult_add_sea has the least value among the models prepared so far 
# Predicting new values 


Model_full = smf.ols('log_Passenger~t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = airlines1).fit()

pred_new  = pd.Series(Model_full.predict(Test))
pred_new

Test["forecasted_Passengers"] = pd.Series(pred_new)




